src/sync/persost-write-transformer.ts

```diff
@@ -0,0 +1,43 @@
+import * as db from '../db/mod';
+import type * as dag from '../dag/mod';
+import type {Hash} from '../hash';
+import type {HashType} from '../db/hash-type';
+
+export type GatheredChunks = ReadonlyMap<Hash, dag.Chunk>;
+
+export class PersistWriteTransformer extends db.Transformer {
+  private readonly _gatheredChunks: GatheredChunks;
+
+  constructor(dagWrite: dag.Write, gatheredChunks: GatheredChunks) {
+    super(dagWrite);
+    this._gatheredChunks = gatheredChunks;
+  }
+
+  protected override shouldForceWrite(h: Hash): boolean {
+    return this._gatheredChunks.has(h);
+  }
+
+  protected override getChunk(hash: Hash): Promise<dag.Chunk | undefined> {
+    const gatheredChunk = this._gatheredChunks.get(hash);
+    return gatheredChunk
+      ? Promise.resolve(gatheredChunk)
+      : super.getChunk(hash);
+  }
+
+  override async transformCommit(
```

Comment:
Just to confirm my understanding: these overrides of `transformCommit` and `transformBTreeNode` ensure we only recurse down until we reach a chunk that the perdag already has.

Reply:
Adding comment:

```ts
// Override to only transform the dag nodes we got from the memdag.
```